rust/cymbal/src/symbol_store/caching.rs

```diff
@@ -0,0 +1,121 @@
+use std::{collections::HashMap, sync::Arc, time::Instant};
+
+use axum::async_trait;
+use tokio::sync::Mutex;
+
+use crate::{
+    error::Error,
+    metric_consts::{
+        STORE_CACHED_BYTES, STORE_CACHE_EVICTIONS, STORE_CACHE_HITS, STORE_CACHE_MISSES,
+        STORE_CACHE_SIZE,
+    },
+};
+
+use super::{SymbolSetRef, SymbolStore};
+
+pub struct CachingStore {
+    inner: Box<dyn SymbolStore>,
+    cache: Mutex<CacheInner>,
+}
+
+impl CachingStore {
+    pub fn new(inner: Box<dyn SymbolStore>, max_bytes: usize) -> Self {
+        metrics::gauge!(STORE_CACHE_SIZE).set(max_bytes as f64);
+        Self {
+            inner,
+            cache: Mutex::new(CacheInner {
+                cached: Default::default(),
+                max_bytes,
+                held_bytes: 0,
+            }),
+        }
+    }
+}
+
+#[async_trait]
+impl SymbolStore for CachingStore {
+    async fn fetch(&self, team_id: i32, r: SymbolSetRef) -> Result<Arc<Vec<u8>>, Error> {
+        let mut cache = self.cache.lock().await;
+
+        if let Some(cached) = cache.get(&r) {
+            metrics::counter!(STORE_CACHE_HITS).increment(1);
+            return Ok(cached);
+        }
+        metrics::counter!(STORE_CACHE_MISSES).increment(1);
+
+        // We hold the cache lock across the underlying fetch, so that if two threads
+        // are racing to fetch the same item, we don't end up doing the request/data transfer twice.
+        let res = self.inner.fetch(team_id, r.clone()).await?;
+
+        cache.insert(r, Arc::clone(&res));
+
+        Ok(res)
+    }
+}
+
+struct CacheInner {
+    // We expect this cache to consist of few, but large, items.
+    cached: HashMap<SymbolSetRef, CachedSymbolSet>,
+    held_bytes: usize,
+    max_bytes: usize,
+}
+
+// TODO - someone smarter than me should replace all this with a proper caching lib,
+// but I'm too lazy, I uhh mean task focused, to go evaluate one right now.
+impl CacheInner {
+    fn insert(&mut self, key: SymbolSetRef, value: Arc<Vec<u8>>) {
+        self.held_bytes += value.len();
+        self.cached.insert(
+            key,
+            CachedSymbolSet {
+                data: value,
+                last_used: Instant::now(),
+            },
+        );
```

Comment:
Should we be worried that a very large source map could cause the rest of the cache to be evicted?

Reply:
I don't think so - we'll be caching in so many places that this is really just so that, when processing a single stack trace, we have all the sourcemaps for all the frames in that stack trace in memory - remember that eventually, once we've processed a stack trace once, we'll never have to lookup a sourcemap to process it again.